### DISCUSSION TITLE
Аренда оборудования

### DISCUSSION THREAD

**Discussion by d-nikolaev** [totumonline/totum-ru-issues-and-discussions#144] [Как это сделать?] [answered]
Есть процессы по передаче в аренду оборудования и их контролю что находится на складе, а что в статусе: арендовано. 
Пытаюсь всё построить от следующей логики Оборудование это главный объект а остальные события кому арендовали и на какой срок второстепенные данные. Создал справочники. 
![Справочники](https://user-images.githubusercontent.com/13610260/170704737-cb8049c6-f57e-4526-8c1a-1a160d7ef970.png)
Дальше попытался построить логику через циклы, но понимаю что не всё ещё изучил про этот инструмент и упёрся. Не получается.
![Циклы](https://user-images.githubusercontent.com/13610260/170705166-06838d70-7858-4d8a-9a2d-e908825c1c81.png)
Нужно регистрировать договор аренды. Договор может быть продлён на новый срок. Самое важное при регистрации нового договора и выбора оборудования в аренду нужно понимать что уже в аренде. 
Вот пример реализации подобной идеи в другой программе.
![Прога](https://user-images.githubusercontent.com/13610260/170706093-87115442-6ea3-456a-82ff-2b7a5b7127b3.png)
Был бы признателен если бы дали направление из каких компонентов реализовать данную задачу.

**Comment by erpic-ru** [comment, marked as answer]
###  [Видео на YouTube](https://youtu.be/FZIbOHPXzXQ)

Этот пример для загрузки: [download link](https://raw.githubusercontent.com/totumonline/ttm-samples/main/samples/totum_sample_11.json.gz.ttm)

Как загрузить: https://github.com/totumonline/ttm-samples

---

Если в договоре аренды один клиент и один тип оборудования, то имеет смысл сделать таблицу циклов и расчетную таблицу в нем (одной будет достаточно). Точно таким же образом, как и показано на последнем скриншоте. 

Чтобы отслеживать то, что в аренде - добавить в таблицу оборудование колонку с датой и передавать в нее дату окончания аренды каждый раз при добавлении новой записи в табличную часть договора аренды по этому оборудованию.

Выборку оборудования (Селект) в договоре аренды сделать по условию, где дата окончания аренды меньше текущей даты.

В колонках таблицы циклов с типом "Ярлык" вывести оборудование, окончание срока аренды + другие нужные поля

**Reply by erpic-ru** [reply]
Вы поля добавляете в колонки. В расчетной таблице в цикле их надо расставить по хедеру, табличной части (колонки) и футеру
https://ru.docs.totum.online/interface#sx-general-view

**Reply by d-nikolaev** [reply]
В договоре клиент один. Но видов оборудования много. 10 видов оборудования в каждом виде набор различных наименования.
![Виды](https://user-images.githubusercontent.com/13610260/170735783-a6aa98af-6d3c-4333-82a3-bbeacf49c4bd.png)

Спасибо про хедер и футер, буду исправлять. Нашел видео про это.

**Reply by erpic-ru** [reply]
Раз нужны 2 табличных части, то да - 2 расчетные таблице в цикле: в одной про оборудование, в другой про сроки

**Reply by ttm-alexeypolunin** [reply]
Тимур предложил рабочий простой вариант, но если похитрее то уточняющие всякие вопросы и я запишу вам видео в вскр:

Правильно ли я понимаю:

1. Клиент берет в аренду позиции со склада
2. Каждая позиция — это отдельная строчка в Оборудование со своим инвентарным номером.
3. У позиций наверное есть тип более общий, те есть Кресло-стул и есть конкретные кресла-стулья №№ 1,2,3...
4. Клиент позицию берет с «даты от» до «дата по»
5. Вам нужно фильтровать какие позиции по типу свободны в нужные клиенту даты
6. Добавить свободную позицию в заказ и учесть ее занятость

Кстати на картинке из другой программы там только одна позиция в договоре с датами или там можно переключится?

По 5 пункту (про интервалы): что происходит если клиент продляет аренду? Есть ли у вас бронь на освободившиеся позиции или вы ставите в заказ только те, которые на склад вернулись?

**Reply by d-nikolaev** [reply]
Спасибо большое за предложенную помощь. Буду вам признателен.
Пункты 1-6, да так именно и есть.
Картинка из другой программы
![GIF 27 05 2022 20-51-31](https://user-images.githubusercontent.com/13610260/170764622-5d601258-e685-4c8d-981d-5e55da09c69b.gif)
Один договор только на одну позицию. Иванов Иван Иванович потребность: костыли, массажер, опора под спину. С ним будут заключены 3 договора на каждую позицию отдельно, со своими сроками а в будущем и пролонгациями. 

**Заметка:** все оборудование выдаётся бесплатно.
Брони нет.
В заказ идут те которые свободны и находятся на складе.

P.S.: В любом случае пока сам всё изучаю и пытаюсь построить то что мне нужно. 
В одно вашем ролике есть вот такая опция для поля
![Сновой строчки](https://user-images.githubusercontent.com/13610260/170766748-e6b4aead-9e88-4ec3-94a0-d7442ffb93a5.png)
Но у себя не могу найти
![Секция](https://user-images.githubusercontent.com/13610260/170766815-a1fa51c4-4f95-4edd-9293-a5374ef74b77.png)

**Reply by ttm-alexeypolunin** [reply]
Она в правой колонке, но только для полей в хедере и футере.

По аренде я понял, в таком варианте все изи, за 10-15 минут я думаю я все объясню.

**Reply by d-nikolaev** [reply]
Понял.

> Она в правой колонке, но только для полей в хедере и футере.

**Reply by ttm-alexeypolunin** [reply]
### [Видео на YouTube](https://youtu.be/FZIbOHPXzXQ)

Чуть позже будет ссылка на возможность загрузки этого примера в вашу базу.

**Reply by ttm-alexeypolunin** [reply]
Этот пример для загрузки: [download link](https://raw.githubusercontent.com/totumonline/ttm-samples/main/samples/totum_sample_11.json.gz.ttm)

Как загрузить: https://github.com/totumonline/ttm-samples

**Reply by d-nikolaev** [reply]
Алексей, спасибо большое за спич, только что посмотрел. Мне выдал Youtube как новинку, я подписан на канал. И после просмотра сразу захожу сюда и вижу, что ещё есть пример для загрузки. Огромное спасибо.
Итак, всё что на видео полностью понятно, понятна логика и пришло понимание правильного использования компонентов Totum.
С Action code, конечно, есть непонимание "что к чему" вернее понимаю, что вы говорите и зачем делаете. Но вот для самостоятельной реализации с чистого листа в другом проекте нужно знать синтаксис и иметь практику.
Ну теперь у меня есть очень интересное домашнее задание, которое я с удовольствием сделаю. 
Обязательно сделаю обзор готового проекта.

**Reply by ttm-alexeypolunin** [reply]
По синтаксису пройдите учебный курс [вот этот раздел](https://ru.docs.totum.online/codes-base-level) и [коды действия](https://ru.docs.totum.online/action-codes-base-level) — это базовые, они на самом деле ключевые.
